Minio/Helper/RequestHelper.cs

```diff
@@ -0,0 +1,162 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using Minio.Exceptions;
+using Minio.Helper;
+using System;
+using System.Text.RegularExpressions;
+
+namespace Minio
+{
+    public class RequestHelper
+    {
+        private static String GetTargetURL(string baseURL, bool secure, string bucketName = null, string region = null, bool usePathStyle = true)
+        {
+            string endPoint = null;
+            string host = baseURL;
+            // For Amazon S3 endpoint, try to fetch location based endpoint.
+            if (s3utils.IsAmazonEndPoint(baseURL))
+            {
+                // Fetch new host based on the bucket location.
+                host = AWSS3Endpoints.Instance.endpoint(region);
+                if (!usePathStyle)
+                {
+                    host = utils.UrlEncode(bucketName) + "." + utils.UrlEncode(host) + "/";
+                }
+            }
+            var scheme = secure ? utils.UrlEncode("https") : utils.UrlEncode("http");
+
+            // This is the actual url pointed to for all HTTP requests
+            endPoint = string.Format("{0}://{1}", scheme, host);
+            
+            return endPoint;
+        }
+        private static Uri GetEndpointURL(string baseURL,bool secure, string bucketName = null, string region = null, bool usePathStyle = true)
+        {
+            if (string.IsNullOrEmpty(baseURL))
+            {
+                throw new InvalidEndpointException("Endpoint cannot be empty.");
+            }
+
+            string endPoint = null;
+            string host = baseURL;
+            string amzHost = baseURL;
+            if ((amzHost.EndsWith(".amazonaws.com", StringComparison.CurrentCultureIgnoreCase))
+                 && !(amzHost.Equals("s3.amazonaws.com", StringComparison.CurrentCultureIgnoreCase)))
+            {
+                throw new InvalidEndpointException(amzHost, "For Amazon S3, host should be \'s3.amazonaws.com\' in endpoint.");
+            }
+            //// For Amazon S3 endpoint, try to fetch location based endpoint.
+            //if (s3utils.IsAmazonEndPoint(baseURL))
+            //{
+            //    // Fetch new host based on the bucket location.
+            //    host = AWSS3Endpoints.Instance.endpoint(region);
+            //    if (!usePathStyle)
+            //    {
+            //        host = utils.UrlEncode(bucketName) + "." + utils.UrlEncode(host) + "/";
+            //    }
+            //}
+            var scheme = secure ? utils.UrlEncode("https") : utils.UrlEncode("http");
```

Comment:
UrlEncode calls seem to be unnecessary here as these strings (http/https) wont be affected by it.